command/agent/template/template.go

```diff
@@ -0,0 +1,194 @@
+func (ts *Server) Run(ctx context.Context, incoming chan string, templates []*ctconfig.TemplateConfig) {
+	latestToken := new(string)
+	ts.logger.Info("starting template server")
+	defer func() {
+		ts.logger.Info("template server stopped")
+		close(ts.DoneCh)
+	}()
+
+	if incoming == nil {
+		panic("incoming channel is nil")
+	}
+
+	// If there are no templates, close the unblockCh
+	if len(templates) == 0 {
+		// nothing to do
+		ts.logger.Info("no templates found")
+		close(ts.unblockCh)
+		return
+	}
+
+	// construct a consul template vault config based the agents vault
+	// configuration
+	var runnerConfig *ctconfig.Config
+	if runnerConfig = newRunnerConfig(ts.config, templates); runnerConfig == nil {
+		ts.logger.Error("template server failed to generate runner config")
+		close(ts.unblockCh)
+		return
+	}
+
+	var err error
+	ts.runner, err = manager.NewRunner(runnerConfig, false)
+	if err != nil {
+		ts.logger.Error("template server failed to create", "error", err)
+		close(ts.unblockCh)
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			ts.runner.StopImmediately()
+			ts.runner = nil
+			return
+
+		case token := <-incoming:
+			if token != *latestToken {
+				ts.logger.Info("template server received new token")
+				ts.runner.Stop()
+				*latestToken = token
+				ctv := ctconfig.Config{
+					Vault: &ctconfig.VaultConfig{
+						Token: latestToken,
+					},
+				}
+				runnerConfig.Merge(&ctv)
+				runnerConfig.Finalize()
+				var runnerErr error
+				ts.runner, runnerErr = manager.NewRunner(runnerConfig, false)
+				if runnerErr != nil {
+					ts.logger.Error("template server failed with new Vault token", "error", runnerErr)
+					// TODO: continue or fail here? Right now we just continue
+					continue
+				} else {
+					go ts.runner.Start()
+				}
+			}
+		case err := <-ts.runner.ErrCh:
+			ts.logger.Error("template server error", "error", err.Error())
+			close(ts.unblockCh)
+			return
+		case <-ts.runner.TemplateRenderedCh():
+			// A template has been rendered, unblock
+			if ts.exitAfterAuth {
+				// if we want to exit after auth, go ahead and shut down the runner
+				ts.runner.Stop()
+			}
+			close(ts.unblockCh)
+		}
+	}
+}
+
+// newRunnerConfig returns a consul-template runner configuration, setting the
+// Vault and Consul configurations based on the clients configs.
+func newRunnerConfig(sc *ServerConfig, templates ctconfig.TemplateConfigs) *ctconfig.Config {
+	// TODO only use default Vault config
+	conf := ctconfig.DefaultConfig()
+	conf.Templates = templates.Copy()
+
+	// Setup the Vault config
+	// Always set these to ensure nothing is picked up from the environment
+	emptyStr := ""
+	conf.Vault.RenewToken = boolPtr(false)
+	conf.Vault.Token = &emptyStr
+	conf.Vault.Address = &sc.VaultConf.Address
+
+	if sc.Namespace != "" {
+		conf.Vault.Namespace = &sc.Namespace
+	}
+
+	conf.Vault.SSL = &ctconfig.SSLConfig{
+		Enabled:    boolPtr(false),
+		Verify:     boolPtr(false),
+		Cert:       &emptyStr,
```

Comment:
Mentioned on the other PR and copying here: A `strPtr()` function seems useful both for consistency and to break the pattern where all of these are being assigned the same pointer. e.g. if somewhere later I do `*conf.Vault.Token = "abc"`, then _all_ of the `&emptyStr` config items would be suddenly be "abc". Breaking the aliasing between these parameters is desirable.

Reply:
Updated in 778721e19 and ca0bb4d1f